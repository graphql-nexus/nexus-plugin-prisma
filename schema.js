module.exports = require('./dist/schema-entrypoint')
