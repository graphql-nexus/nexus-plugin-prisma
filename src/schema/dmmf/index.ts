export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
