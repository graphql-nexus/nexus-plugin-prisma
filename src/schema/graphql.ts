//
// GraphQL root types data & helpers
//

export type RootName = 'Query' | 'Mutation' | 'Subscription'

export const rootNames = {
  Query: 'Query',
  Mutation: 'Mutation',
  Subscription: 'Subscription',
} as const

export const rootNameValues: RootName[] = Object.values(rootNames)

export const isRootName = (x: any): x is RootName => rootNameValues.includes(x)

//
// GraphQL scalar types data & helpers
//

export type ScalarName = 'Int' | 'Float' | 'String' | 'ID' | 'Boolean'

export const scalarsNameValues: ScalarName[] = ['Int', 'Float', 'String', 'ID', 'Boolean']

export const isScalarType = (name: string): boolean => scalarsNameValues.includes(name as any)
