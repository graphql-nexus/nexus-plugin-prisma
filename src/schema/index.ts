import { plugin } from '@nexus/schema'
import { build as buildNexusPrismaTypes, Options, InternalPublicOptions } from './builder'
import { CRUD_NOT_ENABLED } from './warnings'

/**
 * Create a nexus-plugin-prisma instance to be passed into the Nexus plugins array.
 *
 * The nexus plugin extends the Nexus DSL with `t.model` and `t.crud`. These
 * allow you to expose data from your Prisma layer and operations against that
 * data. Once this plugin is installed, you can do things such as:
 *
 * ```ts
 * objectType({
 *   name: 'User',
 *   definition(t) {
 *     t.model.id()
 *     t.model.email()
 *   }
 * })
 *
 * queryType({
 *   definition (t) {
 *      t.crud.user()
 *      t.crud.users({ filtering: true, ordering: true })
 *    }
 * })
 * ```
 *
 * You must ensure the Prisma Client JS has been generated prior as it provides a
 * data representation of the available models and CRUD operations against them.
 *
 * Like Nexus, NexusPrisma has its own Typegen. By default It will be generated
 * synchronously when Nexus runs the plugin, _if_ `process.env.NODE_ENV` is
 * `undefined` or `"development"`. Typegen can be explicitly enabled or disabled
 * via the shouldGenerateArtifacts option. This mirrors Nexus' own typegen
 * approach. This system will change once the Nexus plugin system has first
 * class support for typegen.
 */
export function nexusSchemaPrisma(options?: Options) {
  const allOptions: InternalPublicOptions = options ?? {}
  let wasCrudUsedButDisabled: null | (() => boolean) = null

  return plugin({
    name: 'nexus-plugin-prisma',
    onInstall: (nexusBuilder) => {
      const { types, wasCrudUsedButDisabled: wasCrudUsed } = buildNexusPrismaTypes({
        ...allOptions,
        nexusBuilder,
      })

      wasCrudUsedButDisabled = wasCrudUsed

      return {
        types,
      }
    },
    onBeforeBuild() {
      if (wasCrudUsedButDisabled?.() === true) {
        if (allOptions.initializedByFramework === true) {
          CRUD_NOT_ENABLED.framework()
        } else {
          CRUD_NOT_ENABLED.schema()
        }
      }
    },
  })
}

export { Options } from './builder'
