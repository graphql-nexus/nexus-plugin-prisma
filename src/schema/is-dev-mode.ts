export function isDevMode() {
  return Boolean(!process.env.NODE_ENV || process.env.NODE_ENV === 'development')
}
