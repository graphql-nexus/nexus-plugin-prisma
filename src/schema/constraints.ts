import { DmmfDocument } from './dmmf'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(typeName: string, dmmf: DmmfDocument): string[] {
  const model = dmmf.getModelOrThrow(typeName)

  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0]
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: Record<string, any>,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: Record<string, any>, uniqueIdentifiers: string[]) {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: Record<string, any>, keys: string[]) {
  const output: Record<string, any> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
