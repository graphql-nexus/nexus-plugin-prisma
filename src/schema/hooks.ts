import { colors } from './colors'
import { isDevMode } from './is-dev-mode'

export type OnUnknownFieldName = (info: {
  unknownFieldName: string
  error: Error
  validFieldNames: string[]
  typeName: string
}) => void

export type OnUnknownFieldType = (info: {
  unknownFieldType: string
  error: Error
  typeName: string
  fieldName: string
}) => void

export type OnUnknownArgName = (info: {
  unknownArgNames: string[]
  error: Error
  typeName: string
  fieldName: string
}) => void

export type OnUnknownPrismaModelName = (info: { unknownPrismaModelName: string; error: Error }) => void

type Hooks = OnUnknownFieldType | OnUnknownFieldName | OnUnknownArgName | OnUnknownPrismaModelName | undefined

export function raiseErrorOrTriggerHook<T extends Hooks>(
  hook: T,
  params: Parameters<Exclude<T, undefined>>[0],
  message: string,
  stage: 'build' | 'walk'
) {
  if (stage === 'build') {
    if (isDevMode()) {
      if (hook) {
        ;(hook as any)(params)
      } else {
        const formattedMessage = message
          .split('\n')
          .filter((m) => m.trim().length > 0)
          .map((m) => `${colors.yellow('Warning:')} ${m}`)
          .join('\n')
        console.log(formattedMessage + '\n')
      }
    } else {
      throw new Error(message)
    }
  }
}
