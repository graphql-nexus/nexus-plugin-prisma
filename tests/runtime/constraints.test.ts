import { objectType } from '@nexus/schema'
import { createRuntimeTestContext } from '../__client-test-context'
import { generateSchemaAndTypes } from '../__utils'

let ctx = createRuntimeTestContext()

it('supports nested query with one id field', async () => {
  const datamodel = `
  model Parent {
    idField Int     @id
    child       Child
  }
  
  model Child {
    idField Int    @id
    parent      Parent @relation(fields: [parentId], references: [idField])
  
    parentId Int
  }
`

  const Parent = objectType({
    name: 'Parent',
    definition(t: any) {
      t.model.idField()
      t.model.child()
    },
  })

  const Child = objectType({
    name: 'Child',
    definition(t: any) {
      t.model.idField()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.parent()
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, Parent, Child],
  })

  await dbClient.parent.create({
    data: {
      idField: 1,
      child: {
        create: {
          idField: 1,
        },
      },
    },
  })

  const result = await graphqlClient.request(`{
    parent(where: { idField: 1 }) {
      idField
      child {
        idField
      }
    }
  }`)

  expect(result).toMatchSnapshot()
})

it('supports nested query with compound ids', async () => {
  const datamodel = `
  model Parent {
    idField1 Int
    idField2 Int
    child        Child
  
    @@id([idField1, idField2])
  }
  
  model Child {
    idField1 Int
    idField2 Int
  
    parent        Parent @relation(fields: [parentId1, parentId2], references: [idField1, idField2])
    parentId1 Int
    parentId2 Int
  
    @@id([idField1, idField2])
  }
`

  const Parent = objectType({
    name: 'Parent',
    definition(t: any) {
      t.model.idField1()
      t.model.idField2()
      t.model.child()
    },
  })

  const Child = objectType({
    name: 'Child',
    definition(t: any) {
      t.model.idField1()
      t.model.idField2()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.parent()
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, Parent, Child],
  })

  const { schemaString } = await generateSchemaAndTypes(datamodel, [
    Query,
    Parent,
    Child,
  ])

  console.log(schemaString)

  await dbClient.parent.create({
    data: {
      idField1: 1,
      idField2: 2,
      child: {
        create: {
          idField1: 1,
          idField2: 2,
        },
      },
    },
  })

  const result = await graphqlClient.request(`{
    parent(
      where: {
        idField1_idField2: {
          idField1: 1,
          idField2: 2
        }
      }
    ) {
      idField1
      idField2
      child {
        idField1
        idField2
      }
    }
  }`)

  expect(result).toMatchSnapshot()
})

it('supports nested query without id but one unique', async () => {
  const datamodel = `
  model Parent {
    uniqueField Int     @unique
    child       Child
  }
  
  model Child {
    uniqueField Int    @unique
    parent      Parent @relation(fields: [parentId], references: [uniqueField])
  
    parentId Int
  }
`

  const Parent = objectType({
    name: 'Parent',
    definition(t: any) {
      t.model.uniqueField()
      t.model.child()
    },
  })

  const Child = objectType({
    name: 'Child',
    definition(t: any) {
      t.model.uniqueField()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.parent()
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, Parent, Child],
  })

  await dbClient.parent.create({
    data: {
      uniqueField: 1,
      child: {
        create: {
          uniqueField: 1,
        },
      },
    },
  })

  const result = await graphqlClient.request(`{
    parent(where: { uniqueField: 1 }) {
      uniqueField
      child {
        uniqueField
      }
    }
  }`)

  expect(result).toMatchSnapshot()
})

it('supports nested query without id but multiple uniques', async () => {
  const datamodel = `
  model Parent {
    uniqueField1 Int   @unique
    uniqueField2 Int   @unique
    child        Child
  }
  
  model Child {
    uniqueField1 Int @unique
    uniqueField2 Int @unique
  
    parent   Parent @relation(fields: [parentId], references: uniqueField1)
    parentId Int
  }
`

  const Parent = objectType({
    name: 'Parent',
    definition(t: any) {
      t.model.uniqueField1()
      t.model.uniqueField2()
      t.model.child()
    },
  })

  const Child = objectType({
    name: 'Child',
    definition(t: any) {
      t.model.uniqueField1()
      t.model.uniqueField2()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.parent()
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, Parent, Child],
  })

  await dbClient.parent.create({
    data: {
      uniqueField1: 1,
      uniqueField2: 2,
      child: {
        create: {
          uniqueField1: 1,
          uniqueField2: 2,
        },
      },
    },
  })

  const result = await graphqlClient.request(`{
    parent(where: { uniqueField2: 2 }) {
      uniqueField1
      uniqueField2
      child {
        uniqueField1
        uniqueField2
      }
    }
  }`)

  expect(result).toMatchSnapshot()
})

it('supports nested query without id but compound uniques', async () => {
  const datamodel = `
  model Parent {
    uniqueField1 Int
    uniqueField2 Int
    child        Child
  
    @@unique([uniqueField1, uniqueField2])
  }
  
  model Child {
    uniqueField1 Int
    uniqueField2 Int
  
    parent        Parent @relation(fields: [parentUnique1, parentUnique2], references: [uniqueField1, uniqueField2])
    parentUnique1 Int
    parentUnique2 Int
  
    @@unique([uniqueField1, uniqueField2])
  }
`

  const Parent = objectType({
    name: 'Parent',
    definition(t: any) {
      t.model.uniqueField1()
      t.model.uniqueField2()
      t.model.child()
    },
  })

  const Child = objectType({
    name: 'Child',
    definition(t: any) {
      t.model.uniqueField1()
      t.model.uniqueField2()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.parent()
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, Parent, Child],
  })

  const { schemaString } = await generateSchemaAndTypes(datamodel, [
    Query,
    Parent,
    Child,
  ])

  console.log(schemaString)

  await dbClient.parent.create({
    data: {
      uniqueField1: 1,
      uniqueField2: 2,
      child: {
        create: {
          uniqueField1: 1,
          uniqueField2: 2,
        },
      },
    },
  })

  const result = await graphqlClient.request(`{
    parent(
      where: {
        uniqueField1_uniqueField2: {
          uniqueField1: 1,
          uniqueField2: 2
        }
      }
    ) {
      uniqueField1
      uniqueField2
      child {
        uniqueField1
        uniqueField2
      }
    }
  }`)

  expect(result).toMatchSnapshot()
})
