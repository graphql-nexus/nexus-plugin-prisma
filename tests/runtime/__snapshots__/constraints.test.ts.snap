// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports nested query with compound ids 1`] = `
Object {
  "parent": Object {
    "child": Object {
      "idField1": 1,
      "idField2": 2,
    },
    "idField1": 1,
    "idField2": 2,
  },
}
`;

exports[`supports nested query with one id field 1`] = `
Object {
  "parent": Object {
    "child": Object {
      "idField": 1,
    },
    "idField": 1,
  },
}
`;

exports[`supports nested query without id but compound uniques 1`] = `
Object {
  "parent": Object {
    "child": Object {
      "uniqueField1": 1,
      "uniqueField2": 2,
    },
    "uniqueField1": 1,
    "uniqueField2": 2,
  },
}
`;

exports[`supports nested query without id but multiple uniques 1`] = `
Object {
  "parent": Object {
    "child": Object {
      "uniqueField1": 1,
      "uniqueField2": 2,
    },
    "uniqueField1": 1,
    "uniqueField2": 2,
  },
}
`;

exports[`supports nested query without id but one unique 1`] = `
Object {
  "parent": Object {
    "child": Object {
      "uniqueField": 1,
    },
    "uniqueField": 1,
  },
}
`;
